Content.Server/GameObjects/Components/Disposal/DisposalTubeComponent.cs

```diff
@@ -0,0 +1,380 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Content.Server.Interfaces.GameObjects.Components.Interaction;
+using Content.Shared.GameObjects;
+using Content.Shared.GameObjects.Components.Disposal;
+using Content.Shared.Interfaces;
+using Robust.Server.Console;
+using Robust.Server.GameObjects;
+using Robust.Server.GameObjects.Components.Container;
+using Robust.Server.GameObjects.EntitySystems;
+using Robust.Server.Interfaces.GameObjects;
+using Robust.Shared.GameObjects;
+using Robust.Shared.GameObjects.Components;
+using Robust.Shared.GameObjects.Components.Transform;
+using Robust.Shared.GameObjects.Systems;
+using Robust.Shared.Interfaces.GameObjects;
+using Robust.Shared.Interfaces.Timing;
+using Robust.Shared.IoC;
+using Robust.Shared.Localization;
+using Robust.Shared.Maths;
+using Robust.Shared.Serialization;
+using Robust.Shared.ViewVariables;
+
+namespace Content.Server.GameObjects.Components.Disposal
+{
+    // TODO: Make unanchored pipes pullable
+    public abstract class DisposalTubeComponent : Component, IDisposalTubeComponent, IBreakAct
+    {
+        [Dependency] private readonly IGameTiming _gameTiming = default!;
+
+        private static readonly TimeSpan ClangDelay = TimeSpan.FromSeconds(0.5);
+        private TimeSpan _lastClang;
+
+        private bool _connected;
+        private bool _broken;
+        private string _clangSound;
+
+        /// <summary>
+        ///     Container of entities that are currently inside this tube
+        /// </summary>
+        [ViewVariables]
+        public Container Contents { get; private set; }
+
+        /// <summary>
+        ///     Dictionary of tubes connecting to this one mapped by their direction
+        /// </summary>
+        [ViewVariables]
+        protected Dictionary<Direction, IDisposalTubeComponent> Connected { get; } =
```

Comment:
At least in SS13 disposals tubes don't really form networks and the connections are discovered when something is actively flying through them. This *may* be a good idea because it's less error-prone bookkeeping but your choice.

Reply:
https://github.com/space-wizards/space-station-14/pull/1554